src/viadot/sources/tm1.py

```diff
@@ -0,0 +1,213 @@
+
+        """
+        conn = self.get_connection()
+        return conn.cubes.get_all_names()
+
+    def get_views_names(self) -> list:
+        """Get list of available views in TM1 cube instance.
+
+        Returns:
+            list: List containing available views names.
+
+        """
+        if self.cube is None:
+            msg = "Missing cube name."
+            raise ValidationError(msg)
+        conn = self.get_connection()
+        return conn.views.get_all_names(self.cube)
+
+    def get_dimensions_names(self) -> list:
+        """Get list of available dimensions in TM1 instance.
+
+        Returns:
+            list: List containing available dimensions names.
+
+        """
+        conn = self.get_connection()
+        return conn.dimensions.get_all_names()
+
+    def get_hierarchies_names(self) -> list:
+        """Get list of available hierarchies in TM1 dimension instance.
+
+        Returns:
+            list: List containing available hierarchies names.
+
+        """
+        if self.dimension is None:
+            msg = "Missing dimension name."
+            raise ValidationError(msg)
+
+        conn = self.get_connection()
+        return conn.hierarchies.get_all_names(self.dimension)
+
+    def get_available_elements(self) -> list:
+        """Get list of available elements in TM1 instance.
+
+        Elements are extracted based on hierarchy and dimension.
+
+        Returns:
+            list: List containing available elements names.
+
+        """
+        if (self.dimension or self.hierarchy) is None:
+            msg = "Missing dimension or hierarchy."
+            raise ValidationError(msg)
+
+        conn = self.get_connection()
+        return conn.elements.get_element_names(
+            dimension_name=self.dimension, hierarchy_name=self.hierarchy
+        )
+
+    @add_viadot_metadata_columns
+    def to_df(self, if_empty: Literal["warn", "fail", "skip"] = "skip") -> pd.DataFrame:
+        """Download data into a pandas DataFrame.
+
+            To download the data to the dataframe user needs to specify MDX query
+            or combination of cube and view.
+
+        Args:
+            if_empty (Literal["warn", "fail", "skip"], optional): What to do if output
+            DataFrame is empty. Defaults to "skip".
+
+        Returns:
+            pd.DataFrame: DataFrame with the data.
+
+        Raises:
+            ValidationError: When mdx and cube + view are not specified
+            or when combination of both is specified.
+        """
+        conn = self.get_connection()
+
+        if self.mdx_query is None and (self.cube is None or self.view is None):
+            error_msg = "MDX query or cube and view are required."
+            raise ValidationError(error_msg)
+        if self.mdx_query is not None and (
+            self.cube is not None or self.view is not None
+        ):
+            error_msg = "Specify only one: MDX query or cube and view."
+            raise ValidationError(error_msg)
```

Comment:
These params are provided at init but only validated here; I would remove the params from init and move them here, then this method becomes pure and also it will make sense for these validations to be here.